pennylane/optimize/rotosolve.py

```diff
@@ -12,172 +12,427 @@
+            full_output (bool): whether to return the intermediate minimized energy values from
+                the univariate optimization steps.
+            **kwargs : variable length keyword arguments for the objective function.
+
+        Returns:
+            list [array]: the new variable values :math:`x^{(t+1)}`.
+            If a single arg is provided, list [array] is replaced by array.
+            list [float]: the intermediate energy values, only returned if ``full_output=True``.
+        """
+        x_new, _, *y_output = self.step_and_cost(
+            objective_fn,
+            *args,
+            num_frequencies=num_frequencies,
+            optimizer=optimizer,
+            optimizer_kwargs=optimizer_kwargs,
+            full_output=full_output,
+            **kwargs,
+        )
+        if full_output:
+            return x_new, y_output
+        return x_new
 
-        Updates the parameter :math:`\theta_d` based on Equation 1 in
-        `Ostaszewski et al. (2019) <https://arxiv.org/abs/1905.09692>`_.
+    @staticmethod
+    def _full_reconstruction_equ(fun, num_frequency, fun_at_zero=None):
+        r"""Reconstruct a univariate trigonometric function using trigonometric interpolation.
+        See `Vidal and Theis (2018) <https://arxiv.org/abs/1812.06323>`_ or
+        `Wierichs et al. (2021) <https://arxiv.org/abs/2107.12390>`_.
 
         Args:
-            objective_fn (function): the objective function for optimization. It should take a
-                sequence of the values ``x`` and a list of the gates ``generators`` as inputs, and
-                return a single value.
-            x (Union[Sequence[float], float]): sequence containing the initial values of the
-                variables to be optimized over or a single float with the initial value.
-            d (int): the position in the input sequence ``x`` containing the value to be optimized.
+            fun (callable): the function to reconstruct
+            num_frequency (int): the number of (integer) frequencies present in ``fun``.
+            fun_at_zero (float): The value of ``fun`` at 0. Computed if not provided.
 
         Returns:
-            array: the input sequence ``x`` with the value at position ``d`` optimized.
+            callable: The reconstruction function with ``num_frequency`` frequencies,
+            coinciding with ``fun`` on the same number of points.
         """
-        # helper function for x[d] = theta
-        def insert(x, d, theta):
-            x[d] = theta
-            return x
+        fun_at_zero = float(fun(0.0)) if fun_at_zero is None else fun_at_zero
+        mus = np.arange(1, num_frequency + 1)
+        shifts_pos = 2 * mus * np.pi / (2 * num_frequency + 1)
+        shifts_neg = -shifts_pos[::-1]
+        evals = list(map(fun, shifts_neg)) + [fun_at_zero] + list(map(fun, shifts_pos))
+        shifts = list(shifts_neg) + [0.0] + list(shifts_pos)
```

Comment:
This could potentially be replaced with
```python
np.concatenate([shifts_neg, [0.], shifts_pos])
```
for a speedup, but I don't think this line is much of a bottleneck anyway 😆

Reply:
Nice recommendation! Will do so, not for speedup but niceness :D